include/ensmallen_bits/adam/adamw_update.hpp

```diff
@@ -0,0 +1,114 @@
+ */
+#ifndef ENSMALLEN_ADAM_ADAMW_UPDATE_HPP
+#define ENSMALLEN_ADAM_ADAMW_UPDATE_HPP
+
+#include "adam_update.hpp"
+
+namespace ens {
+
+/**
+ * Adam is an optimizer that computes individual adaptive learning rates for
+ * different parameters from estimates of first and second moments of the
+ * gradients as given in the section 7 of the following paper.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Kingma2014,
+ *   author  = {Diederik P. Kingma and Jimmy Ba},
+ *   title   = {Adam: {A} Method for Stochastic Optimization},
+ *   journal = {CoRR},
+ *   year    = {2014},
+ *   url     = {http://arxiv.org/abs/1412.6980}
+ * }
+ * @endcode
+ */
+class AdamWUpdate
+{
+ public:
+  /**
+   * Construct the Adam update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   * @param weightDecay the rate at which the Update regularises the iterant
+   */
+  AdamWUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
+              const double beta2 = 0.999,
+              const double weightDecay = 0.0005) :
+    optimiser(epsilon, beta1, beta2),
+    weightDecay(weightDecay),
+    iteration(0)
+  {/* Nothing to do.*/ }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    optimiser.Initialize(rows, cols);
+  }
+
+  /**
+   * Update step for AdamW.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    optimiser.Update(iterate, stepSize, gradient);
+    iterate -= weightDecay * iterate;
+  }
+
+  //! Get the value used to initialise the squared gradient parameter.
+  double Epsilon() const { return optimiser.Epsilon(); }
+  //! Modify the value used to initialise the squared gradient parameter.
+  double& Epsilon() { return optimiser.Epsilon(); }
+
+  //! Get the smoothing parameter.
+  double Beta1() const { return optimiser.Beta1(); }
+  //! Modify the smoothing parameter.
+  double& Beta1() { return optimiser.Beta1(); }
+
+  //! Get the second moment coefficient.
+  double Beta2() const { return optimiser.Beta2(); }
+  //! Modify the second moment coefficient.
+  double& Beta2() { return optimiser.Beta2(); }
+
+ private:
+  // The number of iterations.
+  double iteration;
+  // The weight decay rate.
+  double weightDecay;
+  //The AdamUpdate optimser
+  AdamUpdate optimiser;
```

Comment:
To be consistent with the rest of the codebase, do you think, we should go with optimizer or update?